### PR TITLE
Job names can only contain  (alphanumeric|-|_)

### PR DESCRIPTION
**Pull request by jmfayard**
I add an error when running an action:

> The workflow is not valid. .github/workflows/build-mkdocs-website.yml (Line: 13, Col: 3): The identifier 'Build Mkdocs' is invalid. IDs may only contain alphanumeric characters, '_', and '-'. IDs must start with a letter or '_' and and must be less than 100 characters.

### CODE REVIEW

File: library/src/test/kotlin/it/krzeminski/githubactions/yaml/JobsToYamlTest.kt [hunk 1]
```diff
@@ -246,4 +247,16 @@ class JobsToYamlTest : DescribeSpec({
                          |    - name: Some command
                          |      run: echo 'test!'""".trimMargin()
     }
+
+    it("should escape the job name") {
```
Comment (krzema12) [inline library/src/test/kotlin/it/krzeminski/githubactions/yaml/JobsToYamlTest.kt:251]: What do you think about a slightly different approach: failing if the job name contains illegal characters? The reason is that we don't want the user to think that it's fine to use all characters and then silently escape it. It may lead to issues like having two different job names from user's perspective (e. g. "job-1" and "job 1") which would produce the same job name in YAML.

File: library/src/main/kotlin/it/krzeminski/githubactions/domain/Job.kt [hunk 2]
```diff
@@ -8,4 +8,13 @@ data class Job(
     val env: LinkedHashMap<String, String> = linkedMapOf(),
     val condition: String? = null,
     val strategyMatrix: Map<String, List<String>>? = null,
-)
+) {
+
+    // Only chars from [jobNameRegex] are allowed https://docs.github.com/en/actions/using-jobs/using-jobs-in-a-workflow#setting-an-id-for-a-job
+    val escapedName = name.trim().split(jobNameRegex)
```
Comment (krzema12) [inline library/src/main/kotlin/it/krzeminski/githubactions/domain/Job.kt:14]: I suggest moving it to JobsToYaml, where you use `escapedYaml`. The domain class should be as logicless as possible and shouldn't be aware of YAML limitations.

**Review by krzema12 — CHANGES_REQUESTED** [review]

**Comment by krzema12** [comment]
Thanks for approaching it! I created #87 which will prevent such issues to some extent.

File: library/src/test/kotlin/it/krzeminski/githubactions/yaml/JobsToYamlTest.kt (hunk 1 above)
Comment (jmfayard) [reply library/src/test/kotlin/it/krzeminski/githubactions/yaml/JobsToYamlTest.kt:251]: Counter-argument is that when create a new repository on GitHub or a project on Google Cloud, they transform it to whatever internal rules they have, because it works and avoids that the user has to learn about those rules.

I think it's much more likely that someone does a copy-paste and end up having the same exact name ("job" and "job") rather than the scenario "job-1" vs "job 1".

But I can do as you wish.

Comment (krzema12) [reply library/src/test/kotlin/it/krzeminski/githubactions/yaml/JobsToYamlTest.kt:251]: I appreciate the argument of simplifying stuff to the user, that's the point of this library :) However I see a risk of some nasty edge cases if we go the replacement way.

For now I'd like to stay on the safe side and go forward with such mindset: mimic GitHub's error message that you pasted in the PR description, but the added value would be that this library would produce it in build time (generating YAML) rather than runtime (running the workflow). It's more strict and explicit, and lets us loosen it at any time if a good argument appears. The other way round, it would mean a breaking change. As Kotlin scripting evolves, I'm hoping to give the user early feedback in such cases (also: no jobs, no triggers) by failing at script compilation time (also in IDE); it would be possible with a compiler plugin.

Regarding the GitHub behavior - true, however I see an example of prevention to use invalid characters: naming a file in OS. The user eventually needs to learn some basic rules, but it leads to the user coming up with legal names after some time. I'm not an UX expert but I like the fact that some conversion doesn't happen implicitly.

**Comment by jmfayard** [comment]
I updated the logic to fail if the job name is invalid.

File: library/src/test/kotlin/it/krzeminski/githubactions/yaml/JobsToYamlTest.kt [hunk 3]
```diff
@@ -246,4 +249,40 @@ class JobsToYamlTest : DescribeSpec({
...
+        listOf(
+            "job   1",
+            "job1  ",
+            "  job1",
+            "",
+            "-job",
+            "4job",
+            "job()"
+        ).forAll { jobName ->
+            assertThrows<IllegalArgumentException> {
```
Comment (krzema12) [inline library/src/test/kotlin/it/krzeminski/githubactions/yaml/JobsToYamlTest.kt:263]: kotest actually has its own function for this: [`shouldThrow`](https://kotest.io/docs/assertions/exceptions.html). I try to keep stuff platform-agnostic, who knows if this logic hits some other platform than JVM :)

**Review by krzema12 — CHANGES_REQUESTED** [review]
Thanks for the adjustment? One more nitpick and we're ready to merge.

File: library/src/test/kotlin/it/krzeminski/githubactions/yaml/JobsToYamlTest.kt (hunk 3 above)
Comment (jmfayard) [reply library/src/test/kotlin/it/krzeminski/githubactions/yaml/JobsToYamlTest.kt]: Ah yes, that's what I wanted to use but I didn't remember the name.

**Review by krzema12 — APPROVED** [review]
